Pmtud run

fabbione opened the pull request.

fabbione committed:
[rx] kill unused variable

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[stats] allow knet_handle_get_stats to operate in a readlock context

- add global stat mutex lock to protect stats updates
- use global stat mutex lock across all the threads
- fix up some minor bugs:
  - update RX crypto stats only when crypto is enabled
  - update compress and crypto stats in a consistent fashion

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

chrissie-c commented:
I need to go finish for the weekend, I'll leave this comment here so I remember to resume looking at it on Monday.

There appears to be a deadlock in this patch on the link_stats_mutex - more investigation needed.

Reproducable on my 16 node 3 link cluster (but probably will happen elsewhere). do a corosync-cfgtool -s on a running system and it will hang.

Thread 9 (Thread 0x7fdde8367700 (LWP 9182)):
#0  0x00007fddef9a08ed in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007fddef999b09 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00007fddefbb8204 in _link_updown (knet_h=knet_h@entry=0x7fdde866b010, host_id=<optimized out>, link_id=0 '\000', enabled=<optimized out>, 
    connected=connected@entry=0, lock_stats=lock_stats@entry=1) at links.c:64
#3  0x00007fddefbbcb42 in _link_down (dst_host=<optimized out>, dst_link=0x7fddded5eb20, knet_h=0x7fdde866b010) at threads_heartbeat.c:34
#4  _link_down (dst_host=0x7fdddecdc010, dst_link=0x7fddded5eb20, knet_h=0x7fdde866b010) at threads_heartbeat.c:25
#5  _handle_check_each (timed=<optimized out>, dst_link=0x7fddded5eb20, dst_host=0x7fdddecdc010, knet_h=0x7fdde866b010) at threads_heartbeat.c:139
#6  _send_pings (knet_h=0x7fdde866b010, timed=<optimized out>) at threads_heartbeat.c:161
#7  0x00007fddefbbceef in _handle_heartbt_thread (data=0x7fdde866b010) at threads_heartbeat.c:230
#8  0x00007fddef9972de in start_thread () from /lib64/libpthread.so.0
#9  0x00007fddef6c8e83 in clone () from /lib64/libc.so.6

Thread 8 (Thread 0x7fdde8468700 (LWP 9181)):
#0  0x00007fddef9a08ed in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007fddef999b09 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00007fddefbbf34f in _parse_recv_from_links (msg=0x7fdde842fcc0, sockfd=33, knet_h=0x7fdde866b010) at threads_rx.c:739
#3  _handle_recv_from_links (msg=0x7fdde842fcc0, sockfd=33, knet_h=0x7fdde866b010) at threads_rx.c:903
#4  _handle_recv_from_links_thread (data=0x7fdde866b010) at threads_rx.c:964
#5  0x00007fddef9972de in start_thread () from /lib64/libpthread.so.0
#6  0x00007fddef6c8e83 in clone () from /lib64/libc.so.6

Thread 7 (Thread 0x7fdde8569700 (LWP 9180)):
#0  0x00007fddef9a08ed in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007fddef999b09 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00007fddefbc05bc in _dispatch_to_links (knet_h=knet_h@entry=0x7fdde866b010, dst_host=dst_host@entry=0x7fdddecdc010, msg=msg@entry=0x7fdde8524a10, 
    msgs_to_send=msgs_to_send@entry=1) at threads_tx.c:55
#3  0x00007fddefbc1601 in _parse_recv_from_sock (knet_h=knet_h@entry=0x7fdde866b010, inlen=<optimized out>, channel=<optimized out>, 
    channel@entry=0 '\000', is_sync=is_sync@entry=0) at threads_tx.c:589
#4  0x00007fddefbc1d0f in _handle_send_to_links (type=<optimized out>, channel=<optimized out>, sockfd=<optimized out>, msg=0x7fdde8568a70, 
    knet_h=0x7fdde866b010) at threads_tx.c:715
#5  _handle_send_to_links_thread (data=0x7fdde866b010) at threads_tx.c:830
#6  0x00007fddef9972de in start_thread () from /lib64/libpthread.so.0
#7  0x00007fddef6c8e83 in clone () from /lib64/libc.so.6

Thread 6 (Thread 0x7fdde866a700 (LWP 9179)):
#0  0x00007fddef9a08ed in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007fddef999b09 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00007fddefbbc148 in pmtud_reschedule (knet_h=0x7fdde866b010) at threads_common.c:61
#3  get_global_wrlock (knet_h=0x7fdde866b010) at threads_common.c:61
#4  0x00007fddefbbc875 in _handle_dst_link_updates (knet_h=0x7fdde866b010) at threads_dsthandler.c:32
#5  _handle_dst_link_handler_thread (data=0x7fdde866b010) at threads_dsthandler.c:60
#6  0x00007fddef9972de in start_thread () from /lib64/libpthread.so.0
#7  0x00007fddef6c8e83 in clone () from /lib64/libc.so.6

Thread 5 (Thread 0x7fddf06cd700 (LWP 9178)):
#0  0x00007fddef9a08ed in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007fddef999b09 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00007fddefbbd798 in _handle_check_link_pmtud (dst_host=0x7fdddecdc010, dst_link=0x7fddded5eb20, knet_h=0x7fdde866b010) at threads_pmtud.c:198
#3  _handle_check_pmtud (force_run=0, dst_link=0x7fddded5eb20, dst_host=0x7fdddecdc010, knet_h=0x7fdde866b010) at threads_pmtud.c:495
#4  _handle_pmtud_link_thread (data=0x7fdde866b010) at threads_pmtud.c:597
#5  0x00007fddef9972de in start_thread () from /lib64/libpthread.so.0
#6  0x00007fddef6c8e83 in clone () from /lib64/libc.so.6
--Type <RET> for more, q to quit, c to continue without paging--

Thread 4 (Thread 0x7fdde37fe700 (LWP 9177)):
#0  0x00007fddef6c91b7 in epoll_wait () from /lib64/libc.so.6
#1  0x00007fddefbc51ce in _sctp_connect_thread (data=0x7fdde866b010) at transport_sctp.c:651
#2  0x00007fddef9972de in start_thread () from /lib64/libpthread.so.0
#3  0x00007fddef6c8e83 in clone () from /lib64/libc.so.6

Thread 3 (Thread 0x7fdde3fff700 (LWP 9176)):
#0  0x00007fddef6c91b7 in epoll_wait () from /lib64/libc.so.6
#1  0x00007fddefbc4117 in _sctp_listen_thread (data=0x7fdde866b010) at transport_sctp.c:911
#2  0x00007fddef9972de in start_thread () from /lib64/libpthread.so.0
#3  0x00007fddef6c8e83 in clone () from /lib64/libc.so.6

Thread 2 (Thread 0x7fddebf0f700 (LWP 9173)):
#0  0x00007fddef99fe66 in do_futex_wait.constprop () from /lib64/libpthread.so.0
#1  0x00007fddef99ff58 in __new_sem_wait_slow.constprop.0 () from /lib64/libpthread.so.0
#2  0x00007fddf01322c5 in qb_logt_worker_thread () from /lib64/libqb.so.0
#3  0x00007fddef9972de in start_thread () from /lib64/libpthread.so.0
#4  0x00007fddef6c8e83 in clone () from /lib64/libc.so.6

Thread 1 (Thread 0x7fddf0791940 (LWP 9172)):
#0  0x00007fddef9a08ed in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007fddef999b09 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00007fddefbba3c9 in knet_link_get_status (knet_h=0x7fdde866b010, host_id=<optimized out>, link_id=<optimized out>, 
    status=status@entry=0x7ffe0aca0bf0, struct_size=struct_size@entry=1064) at links.c:1175
#3  0x0000557374f854df in totemknet_ifaces_get (knet_context=0x5573766f40d0, status=<optimized out>, iface_count=<optimized out>) at totemknet.c:556
#4  0x0000557374f80a8b in totemsrp_ifaces_get (srp_context=0x557376695340, nodeid=15, interface_id=0x7ffe0acc10e0, interfaces=<optimized out>, 
    interfaces_size=8, status=0x7ffe0acc10d8, iface_count=0x7ffe0acc10d4) at totemsrp.c:1080
#5  0x0000557374f84210 in totempg_ifaces_get (nodeid=<optimized out>, interface_id=<optimized out>, interfaces=<optimized out>, 
    interfaces_size=<optimized out>, status=<optimized out>, iface_count=<optimized out>) at totempg.c:1460
#6  0x0000557374f554a0 in message_handler_req_lib_cfg_get_node_addrs (conn=0x5573790a83a0, msg=<optimized out>) at cfg.c:1038
#7  0x0000557374f6c666 in cs_ipcs_msg_process (c=0x5573790a83a0, data=0x7fddd6bec008, size=<optimized out>) at ipc_glue.c:558
#8  0x00007fddf012b601 in qb_ipcs_dispatch_connection_request () from /lib64/libqb.so.0
#9  0x00007fddf0127da3 in _poll_dispatch_and_take_back_ () from /lib64/libqb.so.0
#10 0x00007fddf012781c in qb_loop_run () from /lib64/libqb.so.0
#11 0x0000557374f4d562 in main (argc=<optimized out>, argv=<optimized out>, envp=<optimized out>) at main.c:1567

fabbione commented:
retest this please

fabbione commented:
@chrissie-c the fix you made actually reverts a covscan fix I did after the original patch was made. We need to figure out if covscan is on crack or we are missing something else.
I am happy to override covscan but we need to make sure we are "right" :)

fabbione committed:
[stats] allow knet_link_get_status to operate in readlock context

- add per link stats mutex
- use per link stats mutex across the board

note: some threads need to lock for a slightly longer period of time than
strictly necessary to avoid reverse-order locking with other mutexes.

Signed-off-by: Christine Caulfield <ccaulfie@redhat.com>
Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione commented:
retest this please

chrissie-c reviewed (dismissed):
ACK in principal, but I'd feel more comfortable if the reporter has had a chance to test it too.

chrissie-c reviewed (approved):
Re-ACK now we've heard back from the reporter. And I have also re-tested it on my 16x3 cluster.
